etcdserver/server.go

```diff
@@ -1611,6 +1615,56 @@ func (s *EtcdServer) RemoveMember(ctx context.Context, id uint64) ([]*membership
 	return s.configure(ctx, cc)
 }
 
+// PromoteMember promotes a learner node to a voting node.
+func (s *EtcdServer) PromoteMember(ctx context.Context, id uint64) ([]*membership.Member, error) {
+	if err := s.checkMembershipOperationPermission(ctx); err != nil {
+		return nil, err
+	}
+
+	if err := s.mayPromoteMember(types.ID(id)); err != nil {
+		return nil, err
+	}
+
+	var memb membership.Member
+	members := s.cluster.Members()
+	isExist := false
+	for _, member := range members {
+		if uint64(member.ID) == id {
+			memb = *member
+			isExist = true
+			break
+		}
+	}
+
+	if !isExist {
```

Comment:
Do we need to validate the promote change before sending to Raft? I looked at other functions such as `AddMember` and `RemoveMember`, they do health checks on cluster, but do not validate the request itself. 

I think it is fine to only validate the request during the later "Apply" stage, i.e., after the request passes through Raft. Add checking here before sending to Raft will cause potential valid request to fail, for example: let's say there is an AddMember request just been proposed to raft, but not applied to `cluster.members` yet. In `PromoteMember` function you will not see that member ID yet, but the request to promote that member might be valid.

Reply:
If we don't check, we may not have enough context for

Reply:
ok